overlord/exportstate/manifest.go

```diff
@@ -0,0 +1,209 @@
+				firstErr = err
+			}
+		}
+	}
+	return firstErr
+}
+
+// exportedFilePath returns the path of an exported file.
+func exportedFilePath(manifest *Manifest, set *ExportSet, exported *ExportedFile) string {
+	snapInstanceName := manifest.SnapInstanceName
+	snapRevision := manifest.SnapRevision.String()
+	if manifest.ExportedForSnapdAsVersion != "" { // Exception for core and host
+		snapInstanceName = "snapd"
+		snapRevision = manifest.ExportedForSnapdAsVersion
+	}
+	return filepath.Join(dirs.ExportDir, snapInstanceName, snapRevision, set.Name, exported.Name)
+}
+
+// exportedFileSourcePath returns the source path that is to be exported.
+func exportedFileSourcePath(manifest *Manifest, set *ExportSet, exported *ExportedFile) string {
+	// Consumer uses host mount namespace
+	if set.ConsumerIsHost {
+		if manifest.SourceIsHost {
+			// host-to-host, no translation needed.
+			return exported.SourcePath
+		}
+		// snap-to-host
+		return filepath.Join(dirs.SnapMountDir, manifest.SnapInstanceName, manifest.SnapRevision.String(), exported.SourcePath)
+	}
+	// Consumer uses snap mount namespace
+	if manifest.SourceIsHost {
+		// host-to-snap, access via hostfs.
+		return filepath.Join("/var/lib/snapd/hostfs", exported.SourcePath)
+	}
+	// snap-to-snap, fixed /snap location inside the snap mount namespace.
+	return filepath.Join("/snap", manifest.SnapInstanceName, manifest.SnapRevision.String(), exported.SourcePath)
+}
+
+// createExportedFile creates an exported file and necessary directories.
+//
+// The function is idempotent.
+//
+// Currently all files are exported as a symbolic link.
+func createExportedFile(manifest *Manifest, set *ExportSet, exported *ExportedFile) error {
+	pathName := exportedFilePath(manifest, set, exported)
+	if err := os.MkdirAll(filepath.Dir(pathName), 0755); err != nil {
+		return err
+	}
+	sourcePath := exportedFileSourcePath(manifest, set, exported)
+	// When there are additional kinds of exported files, switch on the type of
+	// file here. It is possible that some content may be more practical to copy
+	// out of a snap than to link to it.
+	err := os.Symlink(sourcePath, pathName)
+	if err != nil && os.IsExist(err) {
+		if actualTarget, _ := os.Readlink(pathName); actualTarget == sourcePath {
+			err = nil
+		}
+	}
+	return err
+}
+
+// removeExportedFile removes the exported file and prunes any directories.
+func removeExportedFile(manifest *Manifest, set *ExportSet, exported *ExportedFile) error {
+	pathName := exportedFilePath(manifest, set, exported)
+	if err := os.Remove(pathName); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// Chomp three levels up: once for export set name, once for exported
+	// version (or revision) and finally once more for exported snap name (or
+	// snapd). This way we do not need to repeat the special case described in
+	// exportedFilePath.
+	dir := filepath.Dir(pathName)
+	for i := 0; i < 3; i++ {
```

Comment:
I can imagine this going terribly wrong if we are not super careful :-)
Could we instead compute what is the base path where we want to stop, and change the `for` loop condition to `dir != basePath`? (of course, we should also check that `pathName` starts with `basePath`, otherwise this is even more dangerous :-) )